ami/main/models.py

```diff
@@ -2106,45 +2106,123 @@ def root(self):
+        for taxon_id in parents:
+            if taxon_id not in all_parents:
+                taxon_parents = []
+                current_id = taxon_id
+                while current_id in parents:
+                    current_id = parents[current_id]
+                    taxon_parents.append(current_id)
+                all_parents[taxon_id] = taxon_parents
+
+        # Prepare bulk update data
+        bulk_update_data = []
+        for taxon in taxa:
+            taxon_parents = all_parents[taxon.id]
+            parent_taxa = list(taxa.filter(id__in=taxon_parents))
+            taxon_parents = [
+                TaxonParent(
+                    id=taxon.id,
+                    name=taxon.name,
+                    rank=taxon.rank,
+                )
+                for taxon in parent_taxa
+            ]
+            taxon_parents.sort(key=lambda t: t.rank)
 
-        # Update all parents
-        for taxon, parent in parents.items():
-            logging.info(f"Updating parents for {taxon}")
+            bulk_update_data.append(taxon)
 
-            taxon_parents = []
-            while parent:
-                taxon_parents.append(parent)
-                # If this is None, the parent is the root taxon, so we stop here.
-                parent = parents.get(parent)
+        # Perform bulk update
+        # with transaction.atomic():
+        #     self.bulk_update(bulk_update_data, ["parents_json"], batch_size=1000)
+        # There is a bug that causes the bulk update to fail with a custom JSONField
+        # https://code.djangoproject.com/ticket/35167
+        # So we have to update each taxon individually
+        for taxon in bulk_update_data:
+            taxon.save(update_fields=["parents_json"])
 
-            # Convert the taxa to the JSON TaxonParent type
-            taxon_parents = [TaxonParent(id=t.pk, name=t.name, rank=TaxonRank(t.rank)) for t in taxon_parents]
+        logging.info(f"Updated parents for {len(bulk_update_data)} taxa")
 
-            # Sort the parents by rank (achievable because TaxonRank is an ordered enum)
-            taxon_parents.sort(key=lambda t: t.rank)
+    def with_children(self):
+        qs = self.get_queryset()
+        # Add Taxon that are children of this Taxon using parents_json field (not direct_children)
 
-            taxon.parents_json = taxon_parents
-            taxon.save()
+        # example for single taxon:
+        taxon = Taxon.objects.get(pk=1)
+        taxa = Taxon.objects.filter(parents_json__contains=[{"id": taxon.id}])
+        # add them to the queryset
+        qs = qs.annotate(children=models.Subquery(taxa.values("id")))
+        return qs
+
+    def with_occurrence_counts(self) -> models.QuerySet:
+        """
+        Count the number of occurrences for a taxon and all occurrences of the taxon's children.
+        """
+        qs = self.get_queryset()
+
+        # Subquery to get all child taxa IDs
+        import json
+
+        child_taxa = qs.filter(
```

Comment:
Remove experiments until they work. Post woes to the django-pydantic-fields repo